### ISSUE TITLE
sql server backplane for signalr core

### ISSUE THREAD

**Issue by SebaVDPInfra** [aspnet/AspNetCore#5277]
When will signalR sql server backplane be released for dotnet core?

Is there a roadmap of signalR core?

**Comment by sketchau** [comment]
> Sounds great. Thanks all. @anurse yeah I hear you and look forward to seeing what you guys come up with.
> 
> FYI I wound up implementing my own P2P routing system using SignalR connections between the servers and a SQL database just for the servers to "declare" themselves and their IP addresses. It was very straight forward and seems to be working really well. Needs more testing but I'm optimistic. Happy to share results and more specific methods if y'all are interested.

@legistek , it would be great to see your results, if you were willing to share. Using SignalR as a backplane for SignalR seems sensible.

**Comment by legistek** [comment]
@sketchau, @GeXiaoguo I'm not opposed to sharing generally but my particular implementation right now is: a) not tested very much, since we decided to stick with SQL backplaning and ASP.NET Framework for now; and b) not necessarily a general purpose solution as it is built into the main code and assumes our user ID system, etc., so it would be a bit of effort to make it publishable. I was waiting to see what the official project owners came up with for this before thinking about moving to ASP.NET Core and fully implementing something home grown if necessary.

That said what I did was very straightforward and took less than a day to prototype. I'll just explain the basic idea -
- Each load balanced server self-registers with a SQL database during startup. It needs to know its own internal URL for this to work.
- Each server subscribes to a "relay" group of every other registered server
- Each server, whenever it needs to send a message out to one or more groups, publishes that message directly to its client groups, but then also publishes the same message to the "relay" group along with the client group names that should receive it
- Upon receipt of a message through the relay mechanism, a server re-broadcasts the message payload to the group names from the message

In my case each user is identified by Guid, and gets their own Group. So when a server broadcasts messages to other servers for relay it knows the Guids of the users who should get it and includes the Guids with the message. Then each server receiving a message for relay iterates through the users identified in the message and sends the message payload to each group/user.

So really no magic. I just don't know if this is a general purpose solution because I don't know all the different ways people use SignalR.

**Comment by GeXiaoguo** [comment]
@legistek, thanks for sharing the concept. I think it solves a general enough problem. At least, it solves mine problem: two web servers, sharing a single SQL DB, sitting behind a loadbalancer serving a small number of users(<20) with a small number(~1/minute) of messages to publish.

**Comment by maldworth** [comment]
@GeXiaoguo, you might be able to use MassTransit (with RabbitMq). As @anurse mentioned, the Redis implementation is really easy to read. So easy infact that I decided to implement a [SignalR backplane](http://masstransit-project.com/MassTransit/advanced/signalr/) for [MassTransit](https://github.com/masstransit/masstransit/). This does give an alternative option for on prem windows servers, which would be installing RabbitMq Message Broker. Hopefully this is helpful to somebody.

**Comment by GeXiaoguo** [comment]
@maldworth, a practical difficulty in big cooperation is that infrastructure is usually managed by a third party team, or in my case, multiple third party teams. Adding a managed queue like Radis, RabbitMQ is adding a infrastructure. The process cost is too high. That's why there are so many people interested in a SQL Server solution despite its potential scaling problem. Basically every application already has a database. Adding new tables does not need to involve other teams.

**Comment by maldworth** [comment]
Yep, I definitely understand the desire for SQLServer is to use existing infrastructure. I have just found from an infrastructure point of view (who like windows and are familiar), the installers for RabbitMq (rabbitmq server + erlang for windows), is a really easy setup experience. Redis (from what I've found) isn't so easy to setup on windows. So I just wanted to highlight it's an option with slightly less barriers for those Windows + on prem only infrastructure teams.
